src/rootsearch_iterator.jl

```diff
@@ -0,0 +1,134 @@
+
+const BreadthFirstSearch = SearchStrategy{Vector}(push!, shift!)
+const DepthFirstSearch = SearchStrategy{Vector}(push!, pop!)
+
+"""
+    RootSearch{R <: Union{Interval,IntervalBox}, C <: Contractor, CON, T <: Real}
+
+Type implementing the `Base.Iterator` interface to the branch and prune routine.
+Returns the `RootSearchState` at each iteration. Note: Each iteration mutates
+the `RootSearchState`. Use `copy(state::RootSearchState)` to create an
+independent instance if necessary.
+
+# Fields:
+    - `region`: Region in which zeros are searched
+    - `contractor`: Contractor used to determine the status of a region
+    - `strategy`: Strategy determining the order in which regions are processed
+        and the type of container used to store them
+    - `tolerance`: Absolute tolerance of the search
+"""
+struct RootSearch{R <: Union{Interval,IntervalBox}, C <: Contractor, CON, T <: Real}
+    region::R
+    contractor::C
+    strategy::SearchStrategy{CON}
+    tolerance::T
+end
+
+eltype(::Type{RS}) where {R, C, T, CON, RS <: RootSearch{R, C, CON, T}} = RootSearchState{CON{R}, CON{Root{R}}}
+iteratorsize(::Type{RS}) where {RS <: RootSearch} = Base.SizeUnknown()
+
+"""
+    RootSearchState{V, VR}
+
+State type for a `RootSearch` iterator.
+
+The type of the containers used is determinde by the `SearchStrategy` used by
+the `RootSearch`.
+"""
+struct RootSearchState{V, VR}
+    working::V  # Should be a container of the form  CON{T}
+    outputs::VR  # Should ba a container of root of the form CON{Root{T}}
+end
+
+function RootSearchState(rs::RootSearch)
+    return RootSearchState(rs.region, rs.strategy)
+end
+
+function RootSearchState(region::R, strat::SearchStrategy{CON}) where {R <: Union{Interval,IntervalBox}, CON}
+    working = CON{R}()
+    outputs = CON{Root{R}}()
+    strat.store!(working, region)
+    return RootSearchState(working, outputs)
+end
+
+# Currently never reached
+function RootSearchState(region::T) where {T<:Union{Interval,IntervalBox}}
+    working = [region]
+    outputs = Root{T}[]
+
+    sizehint!(working, 1000)
+    sizehint!(outputs, 100)
+
+    RootSearchState(working, outputs)
+end
+
+
+"""
+    copy(state::RootSearchState)
+
+Return an independant copy of `state`. The underlying container type used to
+store regions must support the `deepcopy` function.
+"""
+copy(state::RootSearchState) =
+    RootSearchState(deepcopy(state.working), deepcopy(state.outputs))
+
+
+function start(iter::RootSearch{R, C, CON, T}) where {R, C, CON, T}
+    state = RootSearchState(iter)
+    return state
+end
+
+"""
+    step!(state::RootSearchState, contractor, tolerance)
+
+Progress `state` by treating one of its `working` regions. Note: `state.working`
+is always modified. If a root is found, it is added to `state.outputs`.
+"""
+function step!(state::RootSearchState, contractor, searchstrat, tolerance)
+    X = searchstrat.retrieve!(state.working)
```

Comment:
This line introduces a type instability because `julia` cannot infer the return type of `retrieve!()` for some reason. Three options:
* put the remaining code into a separate function which is then fully inferred
* ignore the issue for now because the heavy lifting is done in `contractor()` which should be fully inferrible
* somehow fix type inferrence for `retrieve!()` (any ideas?)

Reply:
For your convenience; this is how I tested option 1:

```
function step_inner!(X, working, outputs, store!, contractor, tolerance)
    status, output = contractor(X, tolerance)
    if status == :empty
        return nothing
    elseif status == :unique
        store!(outputs, Root(output, :unique))
    elseif diam(output) < tolerance
        store!(outputs, Root(output, :unknown))
    else # branch
        X1, X2 = bisect(X)
        store!(working, X1, X2)
    end
    return nothing
end

"""
    step!(state::RootSearchState, contractor, tolerance)

Progress `state` by treating one of its `working` regions. Note: `state.working`
is always modified. If a root is found, it is added to `state.outputs`.
"""
function step!(state::RootSearchState, contractor, searchstrat, tolerance)
    X = searchstrat.retrieve!(state.working)
    step_inner!(X, state.working, state.outputs, searchstrat.store!, contractor,
                tolerance)
    return nothing
end
```
Here, `step!()` is still unstable but `step_inner!()` is fully inferred.

Reply:
Nice catch! Thinking about it, this is probably why each function has its own type: parametrization on function seems to allow to infer returned type in struct body. Consider the following

```julia
struct A
  func::Function
end

struct B{F}
  func::F
end

a = A(x -> x^2)
b = B(x -> x^2)

test(cont, x) = cont.func(x)
```

In the case of `test(a, 3.2)` the return type can not be inferred, but for `test(b, 3.2)` it can. So this is the way to go.